File: medium/134.js
/**
 * @param {number[]} gas
 * @param {number[]} cost
 * @return {number}
 */
var canCompleteCircuit = function(gas, cost) {
    let idx = 0
    let sum = 0
    let sub_sum = 0
    gas.forEach((e, i) => {
        const v = e - cost[i]
        sum += v
        if (sub_sum < 0) {
            if (v > 0) {
                sub_sum = v
                idx = i
            }
        } else
            sub_sum += v
    })
    return sum < 0 ? -1 : idx
};

File: medium/134.java
/**
    # 134. Gas Station
 */
class Solution {
    public int canCompleteCircuit(int[] gas, int[] cost) {
        final int n = gas.length;
        int earn[] = new int[n];
        
        for(int i = 0; i < n; ++i)
            earn[i] = gas[i] - cost[i];

        int ls = 0; // local sum
        int ts = 0; // total sum
        int st = 0; // start

        for(int i = 0; i < n; ++i){
            ls += earn[i];
            if(ls < earn[i]) {
                ls = earn[i];
                st = i;
            }

            ts += earn[i];
        }

        return ts < 0 ? -1 : st;
    }
}
